linera-base/src/task.rs

```diff
@@ -0,0 +1,69 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+/*!
+Abstractions over tasks that can be used natively or on the Web.
+ */
+
+use std::future::Future;
+
+#[cfg(not(web))]
+mod implementation {
+    use super::*;
+
+    /// The type of errors that can result from awaiting a task to completion.
+    pub type Error = tokio::task::JoinError;
+    /// The type of a future awaiting another task.
+    pub type NonBlockingFuture<R> = tokio::task::JoinHandle<R>;
+    /// The type of a future awaiting another thread.
+    pub type BlockingFuture<R> = tokio::task::JoinHandle<R>;
+
+    /// Spawns a new task, potentially on the current thread.
+    pub fn spawn<F: Future<Output: Send> + Send + 'static>(
+        future: F,
+    ) -> NonBlockingFuture<F::Output> {
+        tokio::task::spawn(future)
+    }
+
+    /// Spawns a blocking task on a new thread.
+    pub fn spawn_blocking<R: Send + 'static, F: FnOnce() -> R + Send + 'static>(
+        future: F,
+    ) -> BlockingFuture<R> {
+        tokio::task::spawn_blocking(future)
+    }
+}
+
+#[cfg(web)]
+mod implementation {
+    use futures::channel::oneshot;
+
+    use super::*;
+
+    /// The type of errors that can result from awaiting a task to completion.
+    pub type Error = oneshot::Canceled;
+    /// The type of a future awaiting another task.
+    pub type NonblockingFuture<R> = oneshot::Receiver<R>;
+    /// The type of a future awaiting another thread.
+    pub type BlockingFuture<R> = oneshot::Receiver<R>;
+
+    /// Spawns a new task, potentially on the current thread.
+    pub fn spawn<F: Future + 'static>(future: F) -> NonblockingFuture<F::Output> {
+        let (send, recv) = oneshot::channel();
+        wasm_bindgen_futures::spawn_local(async {
+            let _ = send.send(future.await);
+        });
+        recv
+    }
+
+    /// Spawns a blocking task on a new thread.
+    /// XXX: not currently implemented: this stub implementation runs the future as a
```

Comment:
Is there an issue for it? Then we could say `TODO(#…)` here.

Reply:
Unfortunately I think when we do create Web workers we're going to need to remove this API entirely.  Web workers just don't really work like `spawn` at all.

Reply:
It could still say `TODO` instead of `XXX`?

Reply:
I've made an issue for it: https://github.com/linera-io/linera-protocol/issues/2399.